dialogue-core/src/main/java/com/palantir/dialogue/core/BaseUrl.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.dialogue.core;
+
+import com.palantir.dialogue.Endpoint;
+import com.palantir.dialogue.Request;
+import com.palantir.dialogue.UrlBuilder;
+import java.net.URL;
+
+/** Convenience utility around {@link UrlBuilder}. */
```

Comment:
Why not add the render method to UrlBuilder?

Reply:
I’m worried about introducing bugs that mutate the urlBuilder state

Reply:
Method on the builder is probably fine but it adds logic to the generator target rather than the runtime, and we still risk mutation

Reply:
Sure...it just seems strange to fully encapsulate one utility class inside another. After this change UrlBuilder wouldn't need to be public, right?

Reply:
The generated Endpoint objects have a method that consumes a UrlBuilder. I think that object should become an interface to further reduce logic in the generator target API, but that can occur separately.